step-capstone/src/components/Trip.js

```diff
@@ -112,78 +21,59 @@ export default class Trip extends React.Component {
         this.handleAddItem = this.handleAddItem.bind(this);
     }
 
-    fetchData() {
-        // TODO: fetch from datastore
-        return testData;
-    }
-
     componentDidMount() {
-        let data = this.fetchData();
-        this.setState({
-            items: data
-        })
+      let travelObjectList = [];
+      this.context.getData(this.state.reference)
+      .then(data => {
+        data.data().travelObjects.forEach(travelObject => {
+          travelObjectList.push(travelObject)
+        });
+        this.setState({items : travelObjectList});
+      })
+      .catch(error => {console.log("Error Getting Trip Data")});
     }
 
-    handleRemoveItem(id) {
+    handleRemoveItem(data) {
+      this.context.deleteTravelObject(this.state.reference, data)
+      .then(() => {
         this.setState({
-            items: this.state.items.filter((item) => item.id !== id)
+          items: this.state.items.filter((item) => item.id !== data.id)
         });
+      })
+      .catch(error => console.log("Error Removing Item"));
     }
 
     handleEditItem(data) {
-        this.setState({
-            items: this.state.items.map((item) => {
-                if (item.id !== data.id) {
-                    return item;
-                } else {
-                    return data
-                }
-            })
-        })
+      this.setState({
+          items: this.state.items.map((item) => {
+              if (item.id === data.id) {
+                let success = this.context.editTravelObject(this.state.reference, item, data)
+                  .then(() => {
+                    return true;
+                  })
+                  .catch(error => {
+                    console.log("Error Editing Item");
+                    return false; 
+                  });
+                return (success ? data : item);
```

Comment:
Is it guaranteed that the editing has been complete at this point? It seems possible that success might be undefined here. I thought only things happening in the .then() were guaranteed to happen when the process is finished, though I could be wrong.

In that case, you could probably eliminate the success variable and instead of returning a boolean, return data when complete and item if error occurs.

Reply:
you are right. i am not entirely sure how to do this correctly - originally I had it the way you are talking about, but that wasnt working either, as anything returned within a then or catch seems to be in the wrong scope - when I return from there it does not return to the .map function. I think I can rework it to use a forEach instead and then just set the item equal to the promise chain like I do with success here but open to ideas here.

Reply:
I think `forEach` and `map` are kind of similar. Current code looks alright to me.